plugins/incremental-ingestion-backend/src/engine/IncrementalIngestionEngine.ts

```diff
@@ -0,0 +1,286 @@
+      { minutes: 1 },
+      { minutes: 5 },
+      { minutes: 30 },
+      { hours: 3 },
+    ];
+  }
+
+  async taskFn(signal: AbortSignal) {
+    try {
+      this.options.logger.debug('Begin tick');
+      await this.handleNextAction(signal);
+    } catch (error) {
+      this.options.logger.error(`${error}`);
+      throw error;
+    } finally {
+      this.options.logger.debug('End tick');
+    }
+  }
+
+  async handleNextAction(signal: AbortSignal) {
+    await this.options.ready;
+
+    const result = await this.getCurrentAction();
+    if (result) {
+      const { ingestionId, nextActionAt, nextAction, attempts } = result;
+
+      switch (nextAction) {
+        case 'rest':
+          if (Date.now() > nextActionAt) {
+            await this.manager.clearFinishedIngestions(
+              this.options.provider.getProviderName(),
+            );
+            this.options.logger.info(
+              `incremental-engine: Ingestion ${ingestionId} rest period complete. Ingestion will start again`,
+            );
+
+            await this.manager.setProviderComplete(ingestionId);
+          } else {
+            this.options.logger.info(
+              `incremental-engine: Ingestion '${ingestionId}' rest period continuing`,
+            );
+          }
+          break;
+        case 'ingest':
+          try {
+            await this.manager.setProviderBursting(ingestionId);
+            const done = await this.ingestOneBurst(ingestionId, signal);
+            if (done) {
+              this.options.logger.info(
+                `incremental-engine: Ingestion '${ingestionId}' complete, transitioning to rest period of ${this.restLength.toHuman()}`,
+              );
+              await this.manager.setProviderResting(
+                ingestionId,
+                this.restLength,
+              );
+            } else {
+              await this.manager.setProviderInterstitial(ingestionId);
+              this.options.logger.info(
+                `incremental-engine: Ingestion '${ingestionId}' continuing`,
+              );
+            }
+          } catch (error) {
+            if (
+              (error as Error).message &&
+              (error as Error).message === 'CANCEL'
+            ) {
+              this.options.logger.info(
+                `incremental-engine: Ingestion '${ingestionId}' canceled`,
+              );
+              await this.manager.setProviderCanceling(
+                ingestionId,
+                (error as Error).message,
+              );
+            } else {
+              const currentBackoff = Duration.fromObject(
+                this.backoff[Math.min(this.backoff.length - 1, attempts)],
+              );
+
+              const backoffLength = currentBackoff.as('milliseconds');
+              this.options.logger.error(error);
+
+              const truncatedError =
+                typeof error === 'string'
+                  ? error.substring(0, 700)
+                  : `${error}`;
```

Comment:
`@backstage/errors` has a `stringifyError` that might be useful here

Reply:
Looks good, but it doesn't truncate really long errors the way ours does. If it could truncate them to prevent enormous DB entries, then it would be a good one to use.

We should probably talk about improving `stringifyError` to properly handle extremely large error stacks.

Reply:
Yeah I meant to do like `stringifyError(error).substring(0, 700)` unconditionally. Your existing code doesn't truncate either, if error is NOT a string and its toString() returns something longer than 700 characters.

Reply:
Ah, I see what you mean. Good idea. Done!